File: 2023/cpp/day03a/day03a.cpp
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <map>
#include <sstream>
#include <utility>

typedef std::pair<int, int> Position;

void usage(const std::string& progname) {
    std::cout << "usage: " << progname << " <file>\n";
    std::exit(1);
}

std::map<Position, std::string> build_numbers(const std::string& contents) {
    std::istringstream iss(contents);
    std::string line;
    std::map<Position, std::string> number_locs;
    int row = 0;
    int col;
    char ch;
    bool scanning_number = false;
    Position current_pos;
    std::string number;
    while ( std::getline(iss, line) ) {
        for ( col = 0; col < line.size(); col++ ) {
            ch = line[col];
            if ( scanning_number ) {
                if ( std::isdigit(ch) ) {
                    number.push_back(ch);
                } else {
                    number_locs.insert({current_pos, number});
                    scanning_number = false;
                    number.clear();
                }
            } else {
                if ( std::isdigit(ch) ) {
                    number.push_back(ch);
                    current_pos = std::make_pair(row, col);
                    scanning_number = true;
                }
            }
        }
        if ( scanning_number ) {
            number_locs.insert({current_pos, number});
            scanning_number = false;
            number.clear();
        }
        row++;
    }
    return number_locs;
}

std::map<Position, char> build_gears(const std::string& contents) {
    std::istringstream iss(contents);
    std::string line;
    std::map<Position, char> gear_locs;
    int row = 0;
    int col;
    char ch;
    while ( std::getline(iss, line) ) {
        for ( col = 0; col < line.size(); col++ ) {
            ch = line[col];
            if ( !std::isdigit(ch) && ch != '.' ) {
                gear_locs.insert({std::make_pair(row, col), ch});
            }
        }
        row++;
    }
    return gear_locs;
}

uint32_t check_gears(const std::map<Position, std::string>& number_locs, const std::map<Position, char>& gear_locs) {
    uint32_t result = 0;
    for ( auto numlocp = number_locs.begin(); numlocp != number_locs.end(); numlocp++ ) {
        int adjacent_count = 0;
        int number_col_start = numlocp->first.second;
        int number_col_end = numlocp->first.second + numlocp->second.size();
        for ( int number_col = number_col_start; number_col < number_col_end; number_col++ ) {
            for ( int delta_row = -1; delta_row <= 1; delta_row ++ ) {
                int adjacent_row = numlocp->first.first + delta_row;
                for ( int delta_col = -1; delta_col <= 1; delta_col++ ) {
                    int adjacent_col = number_col + delta_col;
                    for ( auto gearlocp = gear_locs.begin(); gearlocp != gear_locs.end(); gearlocp++ ) {
                        if ( adjacent_row == gearlocp->first.first && adjacent_col == gearlocp->first.second ) {
                            adjacent_count++;
                        }
                    }
                }
            }
        }
        if ( adjacent_count != 0 ) {
            result += std::atoi(numlocp->second.c_str());
        }
    }
    return result;
}

uint32_t process(const std::string& contents) {
    std::map<Position, std::string> number_locs = build_numbers(contents);
    std::map<Position, char> gear_locs = build_gears(contents);
    return check_gears(number_locs, gear_locs);
}

int main(int argc, char* argv[]) {
    if ( argc < 2 ) {
        usage(argv[0]);
    }

    std::string filename = argv[1];
    std::ifstream file(filename);
    if ( !file.is_open() ) {
        std::cerr << "read of input file failed\n";
        std::exit(1);
    }

    std::ostringstream buffer;
    buffer << file.rdbuf();
    std::string contents = buffer.str();

    uint32_t result = process(contents);
    std::cout << "result = " << result << "\n";

    return 0;
}

File: 2023/elixir/day03a/day03a.exs
defmodule Day03a do
  @progname "day03a.exs"

  def usage do
    IO.puts("usage: #{@progname} <file>")
    System.halt(1)
  end

  def build_numbers(contents) do
    contents |> String.split("\n", trim: true)
    |> Enum.reduce({0, %{}}, fn (line, {row, number_locs}) ->
      {row + 1, Enum.reduce(String.graphemes(line), {0, false, "", {-1, -1}, number_locs}, fn (ch, {col, scanning_number, number, current_pos, inner_number_locs}) ->
        if scanning_number do
          if String.match?(ch, ~r/\d/) do
            if col + 1 == String.length(line) do
              {col + 1, false, "",  {-1, -1}, Map.put(inner_number_locs, current_pos, number <> ch)}
            else
              {col + 1, true, number <> ch, current_pos, inner_number_locs}
            end
          else
            {col + 1, false, "", {-1, -1}, Map.put(inner_number_locs, current_pos, number)}
          end
        else
          if String.match?(ch, ~r/\d/) do
            {col + 1, true, ch, {row, col}, inner_number_locs}
          else
            {col + 1, false, "", current_pos, inner_number_locs}
          end
        end
      end)
      |> elem(4)}
    end)
    |> elem(1)
  end

  def build_gears(contents) do
    contents |> String.split("\n", trim: true)
    |> Enum.reduce({0, %{}}, fn (line, {row, gear_locs}) ->
      {row + 1, Enum.reduce(String.graphemes(line), {0, gear_locs}, fn (ch, {col, inner_gear_locs}) ->
        if !String.match?(ch, ~r/\d/) && ch != "." do
          {col + 1, Map.put(inner_gear_locs, {row, col}, ch)}
        else
          {col + 1, inner_gear_locs}
        end
      end)
      |> elem(1)}
    end)
    |> elem(1)
  end

  def check_gears(number_locs, gear_locs) do
    Map.keys(number_locs) |> Enum.reduce(0, fn (number_loc, result) ->
      number_col_begin = elem(number_loc, 1)
      number_col_end = elem(number_loc, 1) + String.length(Map.get(number_locs, number_loc)) - 1
      found_adjacent = number_col_begin..number_col_end |> Enum.reduce_while(false, fn (number_col, false) ->
        found_adjacent = [{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}] |> Enum.reduce_while(false, fn (neighbor, _) ->
          {adjacent_row, adjacent_col} = {elem(number_loc, 0) + elem(neighbor, 0), number_col + elem(neighbor, 1)}
          found_adjacent = Map.keys(gear_locs) |> Enum.reduce_while(false, fn (gear_loc, _) ->
            if adjacent_row == elem(gear_loc, 0) && adjacent_col == elem(gear_loc, 1) do
              {:halt,  true}
            else
              {:cont, false}
            end
          end)
          if found_adjacent do
            {:halt, true}
          else
            {:cont, false}
          end
        end)
        if found_adjacent do
          {:halt, true}
        else
          {:cont, false}
        end
      end)
      if found_adjacent do
        result + String.to_integer(Map.get(number_locs, number_loc))
      else
        result
      end
    end)
  end

  def process(contents) do
    number_locs = build_numbers(contents)
    gear_locs = build_gears(contents)
    check_gears(number_locs, gear_locs)
  end

  def main do
    if length(System.argv()) < 1 do
      usage()
    end

    filename = hd(System.argv())
    {_, contents} = File.read(filename)
    result = process(contents)
    IO.puts("result = #{result}")
  end

end

Day03a.main()
